chainindex/api.go

```diff
@@ -0,0 +1,167 @@
+			// No tipsets found at this height
+			return 0, 0, nil
+		}
+		return 0, 0, xerrors.Errorf("failed to query tipset counts: %w", err)
+	}
+
+	return revertedCount, nonRevertedCount, nil
+}
+
+func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.ChainEpoch, backfill bool) (*types.IndexValidation, error) {
+	if !si.started {
+		return nil, xerrors.Errorf("ChainValidateIndex can only be called after the indexer has been started")
+	}
+	si.writerLk.Lock()
+	defer si.writerLk.Unlock()
+
+	if err := si.sanityCheckBackfillEpoch(ctx, epoch); err != nil {
+		return nil, err
+	}
+
+	var isIndexEmpty bool
+	err := si.isIndexEmptyStmt.QueryRowContext(ctx).Scan(&isIndexEmpty)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to check if index is empty: %w", err)
+	}
+
+	// fetch the tipset at the given epoch on the canonical chain
+	expectedTs, err := si.cs.GetTipsetByHeight(ctx, epoch, nil, true)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset by height: %w", err)
+	}
+	expectedTsKeyCid, err := expectedTs.Key().Cid()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// Canonical chain has a null round at the epoch -> return if index is empty otherwise validate
+	if expectedTs.Height() != epoch { // Canonical chain has a null round at the epoch
+		if isIndexEmpty {
+			return nil, nil
+		}
+		return si.validateNullRound(ctx, epoch)
+	}
+
+	// if the index is empty -> short-circuit and simply backfill if applicable
+	if isIndexEmpty {
+		return si.backfillMissingTipset(ctx, expectedTs, backfill)
+	}
+
+	revertedCount, nonRevertedCount, err := si.getTipsetCountsAtHeight(ctx, epoch)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return si.backfillMissingTipset(ctx, expectedTs, backfill)
+		}
+		return nil, xerrors.Errorf("failed to get tipset counts at height: %w", err)
+	}
+	if revertedCount == 0 && nonRevertedCount == 0 {
+		return si.backfillMissingTipset(ctx, expectedTs, backfill)
+	}
+
+	if revertedCount > 0 && nonRevertedCount == 0 {
+		return nil, xerrors.Errorf("index corruption: height %d only has reverted tipsets", epoch)
+	}
+
+	if nonRevertedCount > 1 {
+		return nil, xerrors.Errorf("index corruption: height %d has multiple non-reverted tipsets", epoch)
+	}
+
+	// fetch the non-reverted tipset at this epoch
+	var indexedTsKeyCidBytes []byte
+	err = si.getNonRevertedTipsetAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&indexedTsKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get non-reverted tipset at height: %w", err)
+	}
+	indexedTsKeyCid, err := cid.Cast(indexedTsKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	if indexedTsKeyCid != expectedTsKeyCid {
+		return nil, xerrors.Errorf("index corruption: non-reverted tipset at height %d has key %s, but canonical chain has %s", epoch, indexedTsKeyCid, expectedTsKeyCid)
+	}
+
+	return &types.IndexValidation{
```

Comment:
@akaladarshi If the  correct tipset is already indexed at this epoch, we should also make sure that the number of messages and events and message_cids in the Index match what we have in the chainstore. You can see how we get this information from the chainstore in the `indexTipset` function. 

Only checking the number of messages, number of events and all the `message_cids` match is enough for now. No need to compare events or entries. We can do it later after we've shipped this.